### PR TITLE
Fixes #9677 - Lint against mod lib;

### LINKED ISSUE #9677: Lint against `mod lib;`
### What it does

Found that some beginners write `mod lib;` in `main.rs` to reference `lib.rs` contents. Definitely not what people actually want to do!

### Lint Name

_No response_

### Category

suspicious

### Advantage

_No response_

### Drawbacks

_No response_

### Example

```rust
// main.rs
mod lib;
```

Could be written as:

```rust
use this_package as lib;
```

### PR DESCRIPTION
**Pull request by st3fan**
> Warning: Noob in action. This is my first attempt to work on #9677 - work in progress. There is a lot to fix and clean up here.

-- 

Thank you for making Clippy better!

We're collecting our changelog from pull request descriptions.
If your PR only includes internal changes, you can just write
`changelog: none`. Otherwise, please write a short comment
explaining your change.

It's also helpful for us that the lint name is put within backticks (`` ` ` ``),
and then encapsulated by square brackets (`[]`), for example:
```
changelog: [`lint_name`]: your change
```

If your PR fixes an issue, you can add `fixes #issue_number` into this
PR description. This way the issue will be automatically closed when
your PR is merged.

If you added a new lint, here's a checklist for things that will be
checked during review or continuous integration.

- \[ ] Followed [lint naming conventions][lint_naming]
- \[ ] Added passing UI tests (including committed `.stderr` file)
- \[ ] `cargo test` passes locally
- \[ ] Executed `cargo dev update_lints`
- \[ ] Added lint documentation
- \[ ] Run `cargo dev fmt`

[lint_naming]: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

Note that you can skip the above if you are just opening a WIP PR in
order to get feedback.

Delete this line and everything above before opening your PR.

---

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog:

### CODE REVIEW

File: clippy_lints/src/mod_lib.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    /// // example code which does not raise clippy warning
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub MOD_LIB,
+    pedantic,
+    "default lint description"
+}
+declare_lint_pass!(ModLib => [MOD_LIB]);
+
+impl EarlyLintPass for ModLib {
```
Comment (kraktus) [inline clippy_lints/src/mod_lib.rs:27]: Hey, thanks for working on this! I know this PR isn’t ready for review, just wanted to say we’re trying to make all new lints to use ‘LateLintPass’, cheers! Feel free to share any of your issues

Comment (st3fan) [reply clippy_lints/src/mod_lib.rs]: Thanks @kraktus I've updated the code to implement `LateLintPass` instead. Am I going in the right direction here?

Comment (kraktus) [reply clippy_lints/src/mod_lib.rs]: It does! Looks like your lint is not registered yet, using `cargo new_lint` (don’t remember the exact name of the subcommand, find it with `cargo —help` while in clippy repo, will create the skeleton of the lint for you

Comment (llogiq) [reply clippy_lints/src/mod_lib.rs]: Actually `cargo dev update_lint` should suffice in the current situation. The lint exists, but it isn't yet registered.

File: tests/ui/mod_lib/main.rs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#![allow(unused)]
+#![warn(clippy::mod_lib)]
+
+mod lib;
```
Comment (llogiq) [inline tests/ui/mod_lib/main.rs:4]: I think a `mod lib { }` should suffice here, which obviates the need for a secondary file.

**Comment by llogiq** [comment]
Ping! Do you still plan working on this? I think with a rebase and running `cargo dev update_lints` this might mostly be mergeworthy.

@rustbot author

**Comment by Alexendoo** [comment]
cross-posting https://github.com/rust-lang/rust-clippy/issues/9677#issuecomment-1307652711

**Pull request closed without merging**